# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/notification/NotificationUtils.cpp
# Commit: 3daeacd08336
# Full Hash: 3daeacd0833682a474d9bd1adc8817474a7602be
# Author: Kagami Sascha Rosylight <saschanaz@outlook.com>
# Date: 2024-11-08 04:19:40
# Regressor Bug: 1928702
# File Overlap Count: 1
# Description:
#   Bug 1928702 - Part 6: Move NotificationObserver functions to NotificationUtils r=asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D227691
# ==============================================================================

diff -r a497ddb030f6 -r 3daeacd08336 dom/notification/NotificationUtils.cpp
--- a/dom/notification/NotificationUtils.cpp	Thu Nov 07 17:32:05 2024 +0000
+++ b/dom/notification/NotificationUtils.cpp	Thu Nov 07 17:32:05 2024 +0000
@@ -16,6 +16,7 @@
 #include "nsIAlertsService.h"
 #include "nsINotificationStorage.h"
 #include "nsIPermissionManager.h"
+#include "nsIPushService.h"
 #include "nsServiceManagerUtils.h"
 
 namespace mozilla::dom::notification {
@@ -242,4 +243,43 @@
   }
 }
 
+nsresult RemovePermission(nsIPrincipal* aPrincipal) {
+  MOZ_ASSERT(XRE_IsParentProcess());
+  nsCOMPtr<nsIPermissionManager> permissionManager =
+      mozilla::components::PermissionManager::Service();
+  if (!permissionManager) {
+    return NS_ERROR_FAILURE;
+  }
+  permissionManager->RemoveFromPrincipal(aPrincipal, "desktop-notification"_ns);
+  return NS_OK;
+}
+
+nsresult OpenSettings(nsIPrincipal* aPrincipal) {
+  MOZ_ASSERT(XRE_IsParentProcess());
+  nsCOMPtr<nsIObserverService> obs = mozilla::services::GetObserverService();
+  if (!obs) {
+    return NS_ERROR_FAILURE;
+  }
+  // Notify other observers so they can show settings UI.
+  obs->NotifyObservers(aPrincipal, "notifications-open-settings", nullptr);
+  return NS_OK;
+}
+
+nsresult AdjustPushQuota(nsIPrincipal* aPrincipal,
+                         NotificationStatusChange aChange) {
+  nsCOMPtr<nsIPushQuotaManager> pushQuotaManager =
+      do_GetService("@mozilla.org/push/Service;1");
+  if (!pushQuotaManager) {
+    return NS_ERROR_FAILURE;
+  }
+
+  nsAutoCString origin;
+  MOZ_TRY(aPrincipal->GetOrigin(origin));
+
+  if (aChange == NotificationStatusChange::Shown) {
+    return pushQuotaManager->NotificationForOriginShown(origin.get());
+  }
+  return pushQuotaManager->NotificationForOriginClosed(origin.get());
+}
+
 }  // namespace mozilla::dom::notification